Branch-SDK/Branch-SDK/BNCServerInterface.m

```diff
@@ -29,30 +50,51 @@ - (BNCServerResponse *)getRequestSync:(NSDictionary *)params url:(NSString *)url
+    [self postRequestAsync:post url:url andTag:requestTag andLinkData:nil retryNumber:0 retryCount:retryCount log:YES];
 }
 
-- (void)postRequestAsync:(NSDictionary *)post url:(NSString *)url andTag:(NSString *)requestTag log:(BOOL)log {
-    [self postRequestAsync:post url:url andTag:requestTag andLinkData:nil log:log];
+- (void)postRequestAsync:(NSDictionary *)post url:(NSString *)url andTag:(NSString *)requestTag retryCount:(NSInteger)retryCount log:(BOOL)log {
+    [self postRequestAsync:post url:url andTag:requestTag andLinkData:nil retryNumber:0 retryCount:retryCount log:log];
 }
 
-- (void)postRequestAsync:(NSDictionary *)post url:(NSString *)url andTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData {
-    [self postRequestAsync:post url:url andTag:requestTag andLinkData:linkData log:YES];
+- (void)postRequestAsync:(NSDictionary *)post url:(NSString *)url andTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData retryCount:(NSInteger)retryCount {
+    [self postRequestAsync:post url:url andTag:requestTag andLinkData:linkData retryNumber:0 retryCount:retryCount log:YES];
 }
 
 // this is actually a synchronous call; it should NOT be called from the main queue
-- (void)postRequestAsync:(NSDictionary *)post url:(NSString *)url andTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData log:(BOOL)log {
-    BNCServerResponse *serverResponse = [self genericSyncHTTPRequest:[self preparePostRequest:post url:url log:log] withTag:requestTag andLinkData:linkData];
-    if (self.delegate) [self.delegate serverCallback:serverResponse];
+- (void)postRequestAsync:(NSDictionary *)post url:(NSString *)url andTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData retryCount:(NSInteger)retryCount log:(BOOL)log {
+    [self postRequestAsync:post url:url andTag:requestTag andLinkData:linkData retryNumber:0 retryCount:retryCount log:log];
+}
+
+- (void)postRequestAsync:(NSDictionary *)post url:(NSString *)url andTag:(NSString *)requestTag andLinkData:(BNCLinkData *)linkData retryNumber:(NSInteger)retryNumber retryCount:(NSInteger)retryCount log:(BOOL)log {
+    BNCServerResponse *serverResponse = [self genericSyncHTTPRequest:[self preparePostRequest:post url:url retryNumber:retryNumber log:log] withTag:requestTag andLinkData:linkData];
+    
+    NSInteger status = [serverResponse.statusCode integerValue];
+    BOOL isRetryableStatusCode = status >= 500;
+
+    // Retry the request if appropriate
+    if (retryCount > 0 && isRetryableStatusCode) {
+        [NSThread sleepForTimeInterval:[BNCPreferenceHelper getRetryInterval]];
+
+        if (log) {
+            [BNCPreferenceHelper log:FILE_NAME line:LINE_NUM message:@"Replaying request with tag %@", requestTag];
+        }
+
+        [self postRequestAsync:post url:url andTag:requestTag andLinkData:linkData retryNumber:++retryNumber retryCount:--retryCount log:log];
```

Comment:
Hmm, that's true, but I think we'd run into other problems either way if they set their retry count to 99999999999, even with the old method ;)